sycl/doc/EnvironmentVariables.md

```diff
@@ -52,6 +56,14 @@ The device indices are zero-based and are unique only within a backend. Therefor
 
 Additionally, if a sub-device is chosen (via numeric index or wildcard), then an additional layer of partitioning can be specified. In other words, a sub-sub-device can be selected. Like sub-devices, this is done with a period ( `.` ) and a sub-sub-device specifier which is a wildcard symbol ( `*` ) or a numeric index.  Example `ONEAPI_DEVICE_SELECTOR=level_zero:0.*.*` would partition device 0 into sub-devices and then partition each of those into sub-sub-devices. The range of grandchild sub-sub-devices would be the final devices available to the app, neither device 0, nor its child partitions would be in that list. 
 
+Lastly, a filter in the grammar can be thought of as a term in conjuction with an action that is taken on all devices that are selected by the term. The action can be an accept action or a discard action. Based on the action, a filter can be an accept filter or a discard filter.
+The string `<term>` represents an accept filter and the string `!<term>` represents a discard filter. The underlying term is the same but they perform different actions on the matching devices list.
+For example, `!opencl:*` discards all devices of the opencl backend from the list of available devices. The discarding filters, if there are any, must all appear at the end of the selector string. 
+When one or more filters accept a device and one or more filters discard the device, the latter have priority and the device is ultimately not made available to the user. This allows the user to provide selector strings such as `*:gpu;!cuda:*` that accepts all gpu devices except those with a CUDA backend.
+Furthermore, if the value of this environment variable only has discarding filters, an accepting filter that matches all devices, but not sub-devices and sub-sub-devices, will be implicitly included in the 
+environment variable to allow the user to specify only the list of devices that must not be made available. Therefore, `!*:cpu` will accept all devices except those that are of the cpu type and `opencl:*;!*:cpu`
+will accept all devices of the opencl backend exept those that are of the opencl backend and of the cpu type. Finally, it is valid to have no accepting filter match a device but have at least one
+discarding filter match the device. In this case, the device remains unavailable as expected.
```

Comment:
> Finally, it is valid to have no accepting filter match a device but have at least one discarding filter match the device.

This statement seems confusing to me.  We say up above that a term that doesn't match anything is a warning, but I think that is not what you mean here.  I think instead that you mean that it is legal to have a discard action that is not preceded by an accept action.  However, you already say earlier that this is OK:

> if the value of this environment variable only has discarding filters, an accepting filter that matches all devices, but not sub-devices and sub-sub-devices, will be implicitly included in the  environment

Does the statement "Finally, it is valid to have no accepting filter match ..." say anything new?

Reply:
So when I said this I was thinking along the lines of something like opencl:*;!*:gpu if for whatever reason, the user happens to write this. Then a level_zero gpu device would be discarded but it does not appear in any accepting filter so it has a null effect. I thought it would be good to include this just as a remark.

Reply:
> It is legal to have a rejection filter even if it specifies devices have already been omitted by previous filters in the selection string.  Doing so has no effect; the rejected devices are still omitted.

Reply:
Thanks, will update!